app/src/main/java/de/westnordost/streetcomplete/quests/railway_crossing/AddRailwayCrossingBarrier.kt

```diff
@@ -1,21 +1,50 @@
+import de.westnordost.osmapi.map.data.Element
 import de.westnordost.streetcomplete.R
-import de.westnordost.streetcomplete.data.osm.SimpleOverpassQuestType
+import de.westnordost.streetcomplete.data.osm.OsmElementQuestType
 import de.westnordost.streetcomplete.data.osm.changes.StringMapChangesBuilder
+import de.westnordost.streetcomplete.data.osm.download.MapDataWithGeometryHandler
 import de.westnordost.streetcomplete.data.osm.download.OverpassMapDataDao
+import de.westnordost.streetcomplete.data.osm.tql.OverpassQLUtil
+import de.westnordost.streetcomplete.quests.AbstractQuestAnswerFragment
 
-class AddRailwayCrossingBarrier(o: OverpassMapDataDao) : SimpleOverpassQuestType<String>(o) {
+class AddRailwayCrossingBarrier(private val overpassMapDataDao: OverpassMapDataDao) : OsmElementQuestType<RailwayCrossingBarrierAnswer> {
+    override fun createForm(): AbstractQuestAnswerFragment<RailwayCrossingBarrierAnswer> {
+        return AddRailwayCrossingBarrierForm()
+    }
 
-    override val tagFilters = "nodes with railway=level_crossing and !crossing:barrier"
-    override val commitMessage = "Add type of barrier for railway crossing"
-    override val icon = R.drawable.ic_quest_railway
+    override fun download(bbox: BoundingBox, handler: MapDataWithGeometryHandler): Boolean {
+        return overpassMapDataDao.getAndHandleQuota(getOverpassQuery(bbox), handler);
+    }
 
-    override fun getTitle(tags: Map<String, String>) = R.string.quest_railway_crossing_barrier_title
+    override fun isApplicableTo(element: Element): Boolean? {
+        return null
+    }
 
-    override fun createForm() = AddRailwayCrossingBarrierForm()
+    override fun applyAnswerTo(answer: RailwayCrossingBarrierAnswer, changes: StringMapChangesBuilder) {
+        when {
+            answer.barrierType == AddRailwayCrossingBarrierForm.BarrierOption.MISSING -> changes.add("crossing:barrier", "no")
+            answer.barrierType == AddRailwayCrossingBarrierForm.BarrierOption.HALF -> changes.add("crossing:barrier", "half")
+            answer.barrierType == AddRailwayCrossingBarrierForm.BarrierOption.DOUBLE_HALF -> changes.add("crossing:barrier", "double_half")
+            answer.barrierType == AddRailwayCrossingBarrierForm.BarrierOption.FULL -> changes.add("crossing:barrier", "full")
+        }
+    }
 
-    override fun applyAnswerTo(answer: String, changes: StringMapChangesBuilder) {
-        changes.add("crossing:barrier", answer)
+    private fun getOverpassQuery(bbox: BoundingBox): String? {
+        val bboxFilter = OverpassQLUtil.getGlobalOverpassBBox(bbox)
+        return bboxFilter + "way[\"highway\"][\"access\"~\"^private|no\$\"];\n" +
+                "(._;>;);\n" +
+                "node(w) -> .private;\n" +
+                "node[\"railway\"=\"level_crossing\"][!\"crossing:barrier\"]->.crossings;\n" +
+                "(.crossings; - .private;)->.public;\n" +
+                ".public;\n" +
+                "out;"
```

Comment:
Kotlin tip:
Kotlin supports multi line strings with `"""`, so you could write

```kotlin
private fun getOverpassQuery(bbox: BoundingBox) =
   OverpassQLUtil.getGlobalOverpassBBox(bbox) + """
   way["highway"]["access"~"^private|no$"];
   (._;>;);
   node(w) -> .private;
   node["railway"="level_crossing"][!"crossing:barrier"]->.crossings;
   (.crossings; - .private;)->.public;
   .public;
   out;
"""
```
Makes it better readable because you spare all the `"`, `\n`, `+` and esacping `\"`.

Reply:
Also, couldn't you write out .public; or something like this? I thought this is possible

Reply:
`.public out;` works - http://overpass-turbo.eu/s/HDo

Changed.